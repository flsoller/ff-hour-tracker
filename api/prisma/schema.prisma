// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Organization {
  id          String   @id @default(uuid())
  name        String
  description String   @db.VarChar(500)
  createdAt   DateTime @default(now())
  updatedAt   DateTime @updatedAt

  members    Member[]
  activities ActivityType[]

  @@map("organizations")
}

model Member {
  id           String  @id @default(uuid())
  firstName    String
  lastName     String
  emailAddress String  @unique
  isAdmin      Boolean @default(false)

  organization Organization @relation(fields: [orgId], references: [id])
  orgId        String
  timelogs     TimeLog[]

  @@map("members")
}

model ActivityType {
  id           String @id @default(uuid())
  activityName String @unique @db.VarChar(50)
  activityDesc String @db.VarChar(500)

  organization Organization @relation(fields: [orgId], references: [id])
  orgId        String

  timelogs TimeLog[]

  @@map("activity_types")
}

model TimeLog {
  id    String   @id @default(uuid())
  date  DateTime
  hours Int

  activityType   ActivityType @relation(fields: [activityTypeId], references: [id])
  activityTypeId String

  member Member @relation(fields: [id], references: [id])

  @@map("timelogs")
}
