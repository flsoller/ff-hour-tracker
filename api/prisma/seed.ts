import { PrismaClient } from '@prisma/client';
import { hash } from 'bcryptjs';
const prisma = new PrismaClient();
async function main() {
  const hashedPassword = await hash(process.env.ADMIN_USER_PW as string, 12);
  const organization = await prisma.organization.create({
    data: {
      name: 'AdminOrg',
      description: 'Account Management Org',
    },
  });
  await prisma.user.create({
    data: {
      orgId: organization.id,
      emailAddress: process.env.ADMIN_USER_EMAIL as string,
      password: hashedPassword,
      role: 'ADMIN',
      name: 'Admin',
    },
  });
}
main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });
