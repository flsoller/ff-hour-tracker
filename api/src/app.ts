// Package imports
import express, { Application } from 'express';
import cors from 'cors';
import dotenv from 'dotenv';

// Middleware imports
import errorHandler from './middleware/errorHandler';

// Controller imports
import organizations from './controllers/Organizations';
import members from './controllers/Members';
import activities from './controllers/Activities';
import timelog from './controllers/TimeLogs';

// Load env configuration
dotenv.config();

// Initalize express
const app: Application = express();

// Allow cors for development
if (process.env.NODE_ENV === 'development') {
  app.use(cors());
}

// JSON body-parser middleware
app.use(express.json());

// Mount routes
const API_V0 = '/api/v0';
app.use(`${API_V0}/organizations`, organizations);
app.use(`${API_V0}/members`, members);
app.use(`${API_V0}/activities`, activities);
app.use(`${API_V0}/timelog`, timelog);

// Error handling middleware
app.use(errorHandler);

export default app;
