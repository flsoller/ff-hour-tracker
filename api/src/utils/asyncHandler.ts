import { NextFunction, Request, Response } from 'express';

// asyncHandler to avoid repeating try/catch on every async/await controller method.
const asyncHandler =
  (fn: any) => (req: Request, res: Response, next: NextFunction) =>
    Promise.resolve(fn(req, res, next)).catch(next);

export default asyncHandler;
