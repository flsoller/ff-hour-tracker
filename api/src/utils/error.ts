export class ErrorResponse extends Error {
  statusCode: number;
  additionalInfo = {};

  constructor(message: string, statusCode: number, additionalInfo = {}) {
    super(message);
    this.statusCode = statusCode || 500;
    this.additionalInfo = { ...additionalInfo };
  }
}
