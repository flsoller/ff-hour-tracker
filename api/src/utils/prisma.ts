import { PrismaClient } from '@prisma/client';

// Declare prisma on global object to avoid implicit any error
declare global {
  // eslint-disable-next-line no-var
  var prisma: PrismaClient | undefined;
}

/**
 * Instantiate a single instance of prisma
 */
export const prisma =
  global.prisma ||
  new PrismaClient({
    log: ['query'],
  });

if (process.env.NODE_ENV !== 'production') global.prisma = prisma;
