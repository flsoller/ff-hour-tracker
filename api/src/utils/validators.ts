function isTypeOfNumbers(checkTypes: unknown[] = []) {
  return checkTypes.every((item) => Number.isInteger(item));
}

function isValidOrderQuery(query = '') {
  return ['asc', 'desc'].includes(query);
}

export { isTypeOfNumbers, isValidOrderQuery };
