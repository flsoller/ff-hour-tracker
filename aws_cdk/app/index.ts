#!/usr/bin/env node
import 'source-map-support/register';
import * as cdk from 'aws-cdk-lib';
import { HourTrackerApi } from '../lib';

const app = new cdk.App();
new HourTrackerApi(app, 'HourTrackerApi');

app.synth();
