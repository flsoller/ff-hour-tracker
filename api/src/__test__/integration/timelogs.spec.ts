import app from '../../app';
import supertest from 'supertest';
import { prisma } from '../../utils/prisma';
import {
  createOrganizations,
  createMemberForOrganization,
  createUserForOrganization,
  loginTestUser,
} from '../helpers/helpers';
import { Organization } from '@prisma/client';

describe('activities controller', () => {
  const ROUTE = '/api/v0/timelog';
  let organization: Organization;
  let otherOrganization: Organization;
  let token: string;

  beforeEach(async () => {
    [organization, otherOrganization] = await createOrganizations();
    const testUser = await createUserForOrganization(organization.id);
    token = await loginTestUser(app, testUser.emailAddress);
  });

  describe('createTimeLogForDate', () => {
    it('should create a timelog entry for a member', async () => {
      const member = await createMemberForOrganization(
        organization.id,
        'brian@nice-car.com',
        'Brian',
        `O'Connor`,
      );

      const activity = await prisma.activityType.create({
        data: {
          activityName: 'Test Activity',
          activityDesc: 'Detailed description',
          orgId: organization.id,
        },
      });

      const res = await supertest(app)
        .post(ROUTE)
        .auth(token, { type: 'bearer' })
        .send({
          date: new Date(2022, 5, 22),
          hours: 2,
          activityTypeId: activity.id,
          memberId: member.id,
        });

      expect(res.status).toBe(201);
      expect(res.body).toMatchSnapshot();
    });

    it('should throw member not found error if member does not exist for organization', async () => {
      const member = await createMemberForOrganization(otherOrganization.id);

      const activity = await prisma.activityType.create({
        data: {
          activityName: 'Test Activity',
          activityDesc: 'Detailed description',
          orgId: organization.id,
        },
      });

      const res = await supertest(app)
        .post(ROUTE)
        .auth(token, { type: 'bearer' })
        .send({
          date: new Date(2022, 5, 22),
          hours: 2,
          activityTypeId: activity.id,
          memberId: member.id,
        });

      expect(res.status).toBe(400);
      expect(res.body).toStrictEqual({
        error: 'MemberNotFound',
        additionalInfo: {},
      });
    });

    it('should throw activity not found error if it does not exist for organization', async () => {
      const member = await createMemberForOrganization(organization.id);

      const activity = await prisma.activityType.create({
        data: {
          activityName: 'Test Activity',
          activityDesc: 'Detailed description',
          orgId: otherOrganization.id,
        },
      });

      const res = await supertest(app)
        .post(ROUTE)
        .auth(token, { type: 'bearer' })
        .send({
          date: new Date(2022, 5, 22),
          hours: 2,
          activityTypeId: activity.id,
          memberId: member.id,
        });

      expect(res.status).toBe(400);
      expect(res.body).toStrictEqual({
        error: 'ActivityNotFound',
        additionalInfo: {},
      });
    });
  });
});
