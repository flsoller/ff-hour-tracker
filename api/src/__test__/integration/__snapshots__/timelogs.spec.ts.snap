// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activities controller createTimeLogForDate should create a timelog entry for a member 1`] = `
Object {
  "activity": "Test Activity",
  "date": "2022-06-22T00:00:00.000Z",
  "hours": 2,
  "memberName": "Brian O'Connor",
}
`;
